### PR TITLE
Support VB for ImmutableObjectMethodAnalyzer

### LINKED ISSUE #1128: CSharpImmutableObjectMethodAnalyzer does not work when MS.CA.Workspaces is not installed
#### Analyzer package

Microsoft.CodeAnalysis.Analyzers

#### Analyzer

CSharpImmutableObjectMethodAnalyzer 

#### Repro steps

1. Create a new C# project.
2. Install the package `Microsoft.CodeAnalysis.CSharp`, but not `Microsoft.CodeAnalysis.CSharp.Workspaces`
3. Write code that doesn't use the returned value from a method on an immutable object:

   ```c#
   var namespaceNode = SyntaxFactory.NamespaceDeclaration(
       SyntaxFactory.ParseName("Root.MyNamespace1"));
   var classNode = SyntaxFactory.ClassDeclaration("MyClass");

   namespaceNode.AddMembers(classNode);
   ```

#### Expected behavior

The last line reports:

> RS1014 'NamespaceDeclarationSyntax' is immutable and 'AddMembers' will not have any effect on it. Consider using the return value from 'AddMembers'. 

#### Actual behavior

No warning.

---

This is caused by [this line](https://github.com/dotnet/roslyn-analyzers/blob/c457f57/src/Microsoft.CodeAnalysis.Analyzers/CSharp/CSharpImmutableObjectMethodAnalyzer.cs#L60), which means that when symbols from MS.CA.Workspaces are not found, the analyzer does nothing.

### PR DESCRIPTION
**Pull request by Youssef1313**
Write the analyzer using IOperation for easy VB support.

Also fixes #1128

@mavasani I'm not sure whether using IOperation improves performance, doesn't affect it, or make it worse, but if you think this should remain using syntax node actions, let me know and I'll add VB support using syntax node actions.

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx [hunk 1]
```diff
@@ -269,7 +269,7 @@
     <value>'{0}' is immutable and '{1}' will not have any effect on it. Consider using the return value from '{1}'.</value>
   </data>
   <data name="DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationTitle" xml:space="preserve">
-    <value>Do not ignore values returned by methods on immutable objects.</value>
+    <value>Do not ignore values returned by methods on immutable objects</value>
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx:272]: This was reported by RS1031. I don't understand why it wasn't reported earlier, but there is something going wrong.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+            INamedTypeSymbol? type = null;
+            if (invocation.Instance is not null)
+            {
+                type = invocation.Instance.Type as INamedTypeSymbol;
+            }
+            else if (invocation.TargetMethod.IsExtensionMethod && invocation.TargetMethod.Parameters is [{ Type: var parameterType }, ..])
+            {
+                type = parameterType as INamedTypeSymbol;
+            }
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs:99]: @mavasani Does this logic sound correct? I'm surprised there is no property on IInvocationOperation to get this info directly (or at least I couldn't find it)

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx (hunk 1 above)
Comment (Youssef1313) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx:272]: I found the issue and will open an issue.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs:99]: Can you instead use https://github.com/dotnet/roslyn-analyzers/blob/b7bb138809d5a7d31508fe0cd86d59ed4c864764/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L23-L28

**Review by mavasani — APPROVED** [review]

File: src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/ImmutableObjectMethodAnalyzer.cs]: It looks like `GetReceiverType` implementation is incorrect when `beforeConversion` is false.

`firstArg.Type` is always null. I'll use this extension and fix its implementation.

**Pull request merged**